src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -128,6 +133,97 @@ internal partial class Binder
             { WasCompilerGenerated = wasCompilerGenerated };
         }
 
+        private BoundExpression CreateImplicitNewConversion(SyntaxNode syntax, BoundExpression source, Conversion conversion, bool isCast, TypeSymbol destination, DiagnosticBag diagnostics)
+        {
+            var node = (ObjectCreationExpressionSyntax)source.Syntax;
+
+            var arguments = AnalyzedArguments.GetInstance();
+            BindArgumentsAndNames(node.ArgumentList, diagnostics, arguments, allowArglist: true);
+
+            var strippedType = destination.StrippedType();
+            BoundObjectInitializerExpressionBase boundInitializerOpt = node.Initializer != null
+                ? BindInitializerExpression(syntax: node.Initializer, type: strippedType, typeSyntax: syntax, diagnostics)
+                : null;
+
+            BoundExpression operand = source;
+            bool hasErrors;
+            switch (strippedType.TypeKind)
+            {
+                case TypeKind.Array:
+                case TypeKind.Enum:
+                case TypeKind.Delegate:
+                case TypeKind.Interface:
+                    Error(diagnostics, ErrorCode.ERR_BadTargetTypeForNew, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Pointer:
+                    Error(diagnostics, ErrorCode.ERR_UnsafeTypeInObjectCreation, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Dynamic:
+                    Error(diagnostics, ErrorCode.ERR_NoConstructors, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Struct when strippedType.IsTupleType:
+                    Error(diagnostics, ErrorCode.ERR_NewWithTupleTypeSyntax, syntax, strippedType);
+                    hasErrors = true;
+                    break;
+
+                case TypeKind.Struct:
+                case TypeKind.Class:
+                    operand = BindClassCreationExpression(
+                        node,
+                        typeName: strippedType.Name,
+                        typeNode: node,
+                        type: (NamedTypeSymbol)strippedType,
+                        arguments,
+                        diagnostics,
+                        boundInitializerOpt,
+                        forTargetTypedNew: true);
+                    operand.WasCompilerGenerated = true;
+                    hasErrors = operand.HasErrors;
+                    break;
+
+                case TypeKind.TypeParameter:
+                    var typeParameter = (TypeParameterSymbol)strippedType;
+                    if (!typeParameter.IsInstantiable())
+                    {
+                        goto case TypeKind.Array;
```

Comment:
>TypeKind.Array; [](start = 34, length = 15)

Consider inlining the logic from the the array cases here, as the handling of type parameters conceptually is disconnected from arrays (the array case is just used as a coincidental and convenient label).